Feature/improve launcher

lchudinov opened the pull request:
## Improve Zowe Launcher
- Re-use scripts from zowe-install-packaging 
- Standardize logging

Depends on: https://github.com/zowe/zowe-install-packaging/pull/1970

### How to test:

#### When Zowe Launcher runs as part of `run-zowe.sh`
Apply https://github.com/zowe/zowe-install-packaging/pull/1970. Note: `run-zowe.sh` assumes that zowe launcher binary is /bin/zowe_launcher`.

```bash
export INSTANCE_DIR=zowe-instance-dir
export ROOT_DIR=zowe-root-dir
export USE_LAUNCHER=true
"${ROOT_DIR}/bin/internal/run-zowe.sh
```

#### When Zowe Launcher runs as STC
Apply https://github.com/zowe/zowe-install-packaging/pull/1970. Run Zowe with https://github.com/zowe/launcher/blob/9ba3b03c3ed39d47f05faf4edd2bfc0219064aea/samplib/ZWELNCH
or as a quick test:
```bash
export INSTANCE_DIR=zowe-instance-dir
export ROOT_DIR=zowe-root-dir
/path/to/zowe/launcher/bin/zowe_launcher
```

jackjia-ibm commented on src/main.c:
```diff
@@ -480,26 +344,31 @@ static int start_component(zl_comp_t *comp) {
-    fclose(script);
-  } else {
-    fd_map[0] = dup(STDIN_FILENO);
+  snprintf(bin, sizeof(bin), "%s/bin/internal/start-component-with-launcher.sh", getenv("ROOT_DIR"));
```
I think we can use `start-component.sh` directly here. The extra source lines in `start-component-with-launcher.sh` are already included in `prepare-environment.sh` and will be sourced in `start-component.sh`.

1000TurquoisePogs replied:
I see, yes it looks like we can simplify due to https://github.com/zowe/zowe-install-packaging/blob/master/bin/internal/start-component.sh#L54

lchudinov replied:
ok. I changed the source code to use `start-component.sh` directly.

jackjia-ibm commented on samplib/ZWELNCH:
```diff
@@ -34,5 +37,6 @@ RPTOPTS(ON)
 //STDENV   DD  *
-WORKDIR=#zwelnch
+INSTANCE_DIR=#instance_dir
+ROOT_DIR=#root_dir
```
Any possibility to remove `ROOT_DIR` from here?

1000TurquoisePogs replied:
When running as a script it would be located in ROOT_DIR, so it seems it would not be needed.
But as an STC, it seems that zowe launcher runs outside of ROOT_DIR, so it needs to know where it is.
Is that right?

jackjia-ibm replied:
If it's STC, we should have copied ZWELNCH to loadlib. My concern is if we replace ZWESVSTC with ZWELNCH, it will be so different and how should we modify this doc: https://docs.zowe.org/stable/user-guide/configure-zowe-server.html#option-2-starting-zowe-with-a-s-tso-command.

lchudinov replied:
Yes, I see two possibilities:

1. Zowe Launcher launcher gets `ROOT_DIR` by reading and interpreting `${INSTANCE_DIR}/instance.env` file
2. Zowe Launchers executes some script inside ${INSTANCE_DIR} that could print ${ROOT_DIR} into `stdout`

Which one is better @jackjia-ibm ?

jackjia-ibm reviewed (commented):
Thanks Leonty!

I didn't test using launcher as STC yet, so far the test with launching from run-zowe.sh looks good and all services are started. There are some logging issues if you check [STDOUT.txt](https://github.com/zowe/launcher/files/6018225/STDOUT.txt):

- launcher logging doesn't follow standard `<timestamp> <<process>:<pid>> <service-id> <loglevel> (<stack>) <message>` yet
- pid showing after the service is the pid of start-component.sh script, not the the pid of real service. For example, `zss(16844372)` is `ZWE1SV8  STC04582 SWAPPED,WAITING FOR CHILD        ZWESVUSR 1WI       0.06   16844372`.
- some lines are truncated. for example line 57 `2021-02-21 17:58:33 CMSG:  zss(16844372) - Pl`.
- launcher wraps original logging lines like this: `2021-02-21 17:58:33 CMSG:  zss(16844372) - 2021-02-21 17:58:32 <ZWELS:16844372>  INFO (start-component.sh:83) starting component zss ...`

Quick question, I tried `/F ZWELNCH,APPL=STOP(app-server)` but got this error: `RESPONSE=S0W1      IEE342I MODIFY   REJECTED-TASK BUSY`. Does that mean if we want to support `F` command, we have to use launcher as STC?

Do you mind I modify your https://github.com/zowe/zowe-install-packaging/pull/1970 to embed launcher into packaging and installation?

ifakhrutdinov commented:
@lchudinov can you please describe what's being improved so that the code changes can be verified against the new expected behavior?

lchudinov commented:
@jackjia-ibm thanks for giving it a try.

> Quick question, I tried `/F ZWELNCH,APPL=STOP(app-server)` but got this error: `RESPONSE=S0W1 IEE342I MODIFY REJECTED-TASK BUSY`. Does that mean if we want to support `F` command, we have to use launcher as STC?

It looks like so.

> Do you mind I modify your [zowe/zowe-install-packaging#1970](https://github.com/zowe/zowe-install-packaging/pull/1970) to embed launcher into packaging and installation?

Sure, feel free to modify both my PRs.

lchudinov commented:
@ifakhrutdinov The intent of this PR is to correctly reuse the scripts that we already have in zowe-install-packaging.
Please see my comments on https://app.zenhub.com/workspaces/community-5c93e02fa70b456d35b8f0ed/issues/zowe/zlux/590

ifakhrutdinov commented:
@lchudinov thanks. It could be better to have some that information in the description of the pull-request. It would save the reviewers and you time, because there will be more questions in the diff otherwise.

1000TurquoisePogs commented on src/main.c:
```diff
@@ -110,41 +121,63 @@ struct {
   zl_context.config = *cfg;
 
-  if (chdir(zl_context.workdir)) {
+  zl_context.start_mode = argc > 1 ? ZL_START_MODE_PS : ZL_START_MODE_STC;
```
Determining whether to use STC or process by argument count where arguments = list of components

1000TurquoisePogs commented:
@jackjia-ibm 
> launcher logging doesn't follow standard <timestamp> <<process>:<pid>> <service-id> <loglevel> (<stack>) <message> yet
Is that even a good idea for STC? It works for scripts when you can log to a file and view it in an editor of your choice, but when I see people trying to read our logs from a 3270 terminal it does not look easy. Is that expected by sys admins or do they want logs that are under 80 columns? If so, perhaps zowe launcher should not use our log format in order to have short logs.

ifakhrutdinov commented on src/main.c:
```diff
@@ -110,41 +121,63 @@ struct {
   zl_context.config = *cfg;
 
-  if (chdir(zl_context.workdir)) {
+  zl_context.start_mode = argc > 1 ? ZL_START_MODE_PS : ZL_START_MODE_STC;
```
This is a rather hacky way to determine the mode. Is there a better way?

lchudinov replied:
The way exactly does what I wanted to. There are two modes: when the launcher starts from `run-zowe.sh` and when the launcher starts independently. I want two distinguish the way by number of command line arguments. Perhaps `ZL_START_MODE_PS` and `ZL_START_MODE_STC` are not very apt names and should be replaced. @ifakhrutdinov  If you have any suggestions please add a comment.

ifakhrutdinov replied:
I've done some quick research. There seems to be a field in ASCB you could use to distinguish between STC/job/process.

If you follow ASCB->ASCBJBNS, you'll see the job name used with the start command (if I interpreted that correctly).

Let's consider I use the following applications:
* An STC with name `MYSTC`
* A job with name `MYJOB`
* A process `sh test.sh`

||STC|Job|Process|
|-|-|-|-|
|**ASCBJBNS**|MYSTC|INIT|BPXAS|

Even using this method you can check if the value is `BPXAS`, you're a USS process. There may be better way, I'll check with experts and get back to you.

ifakhrutdinov replied:
I've been told that better ways are:
* Checking `OUCBOMVS` in `OUCBMFL3` (see [OUCB](https://www.ibm.com/support/knowledgecenter/en/SSLTBW_2.3.0/com.ibm.zos.v2r3.iead300/OUCB-map.htm))
* `IEECHAIN`

ifakhrutdinov commented on src/main.c:
```diff
@@ -875,17 +744,77 @@ static int send_event(enum zl_event_t event_type, void *event_data) {
     exit(EXIT_FAILURE);
   }
+  
+  char buf[512];
```
Can you use a more descriptive name? `comp_buffer`?

lchudinov replied:
Sure.

ifakhrutdinov commented on src/main.c:
```diff
@@ -875,17 +744,77 @@ static int send_event(enum zl_event_t event_type, void *event_data) {
+  
+  if (zl_context.start_mode == ZL_START_MODE_STC) {
+    if (prepare_workspace()) {
+      return -1;
```
`exit(EXIT_FAILURE)` for consistency.

lchudinov replied:
Sure.

ifakhrutdinov commented on src/main.c:
```diff
@@ -875,17 +744,77 @@ static int send_event(enum zl_event_t event_type, void *event_data) {
+static int prepare_workspace() {
+  char command[PATH_MAX];
+  INFO("about to prepare zowe workspace\n");
+  const char *script = "bin/internal/prepare-workspace-for-launcher.sh";
```
This ties the launcher to the zowe-install-packaging repo. Can we externalize this?

lchudinov replied:
Not sure. How could we pass the script name to the launcher? via environment variable?

jackjia-ibm commented:
@1000TurquoisePogs I think a unified logging format is important, it provides possibility we can convert each line into a JSON logging entry (with correct timestamp, component id, log level, etc) and parse it with other popular log analysis tools. It eventually will help sysadm to search through logs, or use any AI tools to analysis logs, add monitoring by properly identified log level, etc. Maybe our logging format standard is not ideal, I think SDSF.LOG or z/OSMF log could be a better for TN3270 screen. They both support multiple-line log entry and there is rule to determine where a new log entry starts. But changing this will require other components to make same change and become a complicated task.

---

I noticed another issue starting launcher with USS but I didn't double check. I started Zowe on marist with run-zowe.sh+launcher during the weekend, then the nightly build use zowe-stop.sh to cancel the job before starting a new test on nightly. Launcher itself exited but the sub-processes are left running. Maybe the best way is still use launcher as STC.

jackjia-ibm commented:
Thanks Leonty for all the fixes. I did some tests on [build #30](https://wash.zowe.org:8443/view/Zowe/job/zowe-launcher/view/change-requests/job/PR-18/30/), and here is the result:

* starting with run-zowe.sh
  - i tried both C and P command to terminate the job, but both left subprocesses in the memory. seems signals are not sent to subprocesses.
* starting as STC
  - I got error:
    ```
2021-02-27 17:38:37 <ZWELNCH:33556759> ZWESVUSR ERROR command '/var/zowe/runtime/bin/internal/prepare-workspace-for-launcher.sh -c /
var/zowe/instance -r /var/zowe/runtime' ended with code 256                                                                         
    ```
    and found `prepare-workspace-for-launcher.sh` exit code is not set correctly. after add line `exit 0` at the end of this file solved the problem and all services are started
  - found `STDERR` content is not showing in the log
  - tested `F ZWELNCH,APPL=STOP(app-server)` and `F ZWELNCH,APPL=START(app-server)` works!
  - `C ZWELNCH` will leave subprocesses in memory
  - `P ZWELNCH` works well

cleonty commented:
@jackjia-ibm thanks for testing it.
 
> * starting with run-zowe.sh
>   
>   * i tried both C and P command to terminate the job, but both left subprocesses in the memory. seems signals are not sent to subprocesses.

Did you run `zun-zowe.sh` with using `BPXBATSL`?

> * starting as STC
>   
>   * I got error:
>     ```
>     2021-02-27 17:38:37 <ZWELNCH:33556759> ZWESVUSR ERROR command '/var/zowe/runtime/bin/internal/prepare-workspace-for-launcher.sh -c /
>     var/zowe/instance -r /var/zowe/runtime' ended with code 256                                                                         
>     ```
>     
>     
>     and found `prepare-workspace-for-launcher.sh` exit code is not set correctly. after add line `exit 0` at the end of this file solved the problem and all services are started

I added `exit 0`  at the end of `prepare-workspace-for-launcher.sh` in my PR.

>   * found `STDERR` content is not showing in the log

I fixed the issue.

>   * tested `F ZWELNCH,APPL=STOP(app-server)` and `F ZWELNCH,APPL=START(app-server)` works!

Great!

>   * `C ZWELNCH` will leave subprocesses in memory

Please use `STOP` command to stop the STC.

>   * `P ZWELNCH` works well

Great!

lchudinov commented:
@jackjia-ibm thanks for testing it.
 
> * starting with run-zowe.sh
>   
>   * i tried both C and P command to terminate the job, but both left subprocesses in the memory. seems signals are not sent to subprocesses.

Did you run `zun-zowe.sh` with using `BPXBATSL`?

> * starting as STC
>   
>   * I got error:
>     ```
>     2021-02-27 17:38:37 <ZWELNCH:33556759> ZWESVUSR ERROR command '/var/zowe/runtime/bin/internal/prepare-workspace-for-launcher.sh -c /
>     var/zowe/instance -r /var/zowe/runtime' ended with code 256                                                                         
>     ```
>     
>     
>     and found `prepare-workspace-for-launcher.sh` exit code is not set correctly. after add line `exit 0` at the end of this file solved the problem and all services are started

I added `exit 0`  at the end of `prepare-workspace-for-launcher.sh` in my PR.

>   * found `STDERR` content is not showing in the log

I fixed the issue.

>   * tested `F ZWELNCH,APPL=STOP(app-server)` and `F ZWELNCH,APPL=START(app-server)` works!

Great!

>   * `C ZWELNCH` will leave subprocesses in memory

Please use `STOP` command to stop the STC.

>   * `P ZWELNCH` works well

Great!

ifakhrutdinov commented:
@1000TurquoisePogs, @lchudinov, @jackjia-ibm , see my comment in https://github.com/zowe/zlux/issues/590#issuecomment-787923146, let have a discussion and determine the architecture.

jackjia-ibm commented:
> @jackjia-ibm thanks for testing it.
> 
> > * starting with run-zowe.sh
> >   
> >   * i tried both C and P command to terminate the job, but both left subprocesses in the memory. seems signals are not sent to subprocesses.
> 
> Did you run `zun-zowe.sh` with using `BPXBATSL`?
> 

Yes it's started with regular `<instance-dir>/bin/zowe-start.sh`, with current `ZWESVSTC` and `BPXBATSL`.

jackjia-ibm commented:
@lchudinov as mentioned during the meeting, I need some helps on by adding new parameter to `samplib/ZWESLSTC`:
- add `HA_INSTANCE_ID` as parameter and pass along
- these shell scripts take it as `-i` option like `bin/internal/script.sh ... -i "${HA_INSTANCE_ID}"`
  * `bin/internal/start-component.sh`
  * `bin/internal/get-launch-components.sh`
  * `bin/internal/prepare-instance.sh`
- no need to change `bin/internal/read-essential-vars.sh`

Thanks a lot Leonty!

jackjia-ibm reviewed (approved):
This works ok with zowe-install-packaging#2049 test. Thanks a lot!

jackjia-ibm commented:
Will merge and release, so we can update PR2049 with formal version
